File: frontend/src/email-pdf.ts
import { requestJSON } from "./api.js";

export interface EmailPDFReceipt {
  source: { node_id: number; version_id: string; sha256: string; size: number };
  attachment_id: string;
  profile_fingerprint: string;
  binding: { generation_id: string; recipe: { paper: string } };
  output: { pdf_sha256: string; pdf_size: number; pages: number };
}

export function validatePDFReceipt(value: unknown, versionID: string, profile?: string): EmailPDFReceipt {
  const r = value as Partial<EmailPDFReceipt> | null;
  const hash = (v: unknown) => typeof v === "string" && /^[a-f0-9]{64}$/.test(v);
  if (!r || r.source?.version_id !== versionID || !hash(r.source.sha256) || !hash(r.attachment_id) || !hash(r.profile_fingerprint) || (profile && r.profile_fingerprint !== profile) || !hash(r.binding?.generation_id) || !["A4", "Letter"].includes(r.binding?.recipe?.paper ?? "") || !hash(r.output?.pdf_sha256) || !Number.isSafeInteger(r.output?.pdf_size) || (r.output?.pdf_size ?? 0) < 1 || (r.output?.pdf_size ?? Infinity) > 256 * 1024 * 1024 || !Number.isSafeInteger(r.output?.pages) || (r.output?.pages ?? 0) < 1 || (r.output?.pages ?? Infinity) > 1000) {
    throw new Error("The PDF receipt disagrees with the selected email version or recipe.");
  }
  return r as EmailPDFReceipt;
}

export async function retainedEmailPDFs(session: string, versionID: string, signal: AbortSignal): Promise<EmailPDFReceipt[]> {
  const values = await requestJSON<unknown[]>(`/api/v1/email-pdfs/${encodeURIComponent(versionID)}`, session, { signal });
  if (!Array.isArray(values)) throw new Error("Invalid retained PDF response.");
  return values.map((v) => validatePDFReceipt(v, versionID));
}

export async function renderEmailPDF(session: string, versionID: string, paper: string, signal: AbortSignal): Promise<EmailPDFReceipt> {
  const job = await requestJSON<{ job_id: string; version_id: string; profile_fingerprint: string; state: string }>("/api/v1/email-pdfs", session, { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify({ version_id: versionID, paper, consent: true }), signal });
  if (job.version_id !== versionID || !/^[a-f0-9]{64}$/.test(job.profile_fingerprint) || !/^[a-f0-9]{64}$/.test(job.job_id)) throw new Error("The PDF job disagrees with the selected version.");
  while (job.state !== "completed") {
    if (!["queued", "running", "retry_wait"].includes(job.state)) throw new Error(`PDF rendering ended ${job.state}; inspect processing jobs.`);
    await new Promise<void>((resolve, reject) => {
      if (signal.aborted) { reject(signal.reason); return; }
      const abort = () => { clearTimeout(timer); reject(signal.reason); };
      const timer = setTimeout(() => { signal.removeEventListener("abort", abort); resolve(); }, 500);
      signal.addEventListener("abort", abort, { once: true });
    });
    const next = await requestJSON<{ state: string }>(`/api/v1/email-pdf-jobs/${encodeURIComponent(job.job_id)}`, session, { signal });
    job.state = next.state;
  }
  const receipt = await requestJSON<unknown>(`/api/v1/email-pdfs/${encodeURIComponent(versionID)}/${job.profile_fingerprint}`, session, { signal });
  return validatePDFReceipt(receipt, versionID, job.profile_fingerprint);
}

File: internal/client/email_pdf.go
package client

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"go.kenn.io/docbank/document"
	"go.kenn.io/docbank/document/emailpdf"
	"io"
	"net/http"
	"net/url"
)

func (c *Client) RequestEmailPDF(ctx context.Context, input document.EmailPDFRequest) (document.EmailPDFJob, error) {
	var out document.EmailPDFJob
	if !validUUIDv4(input.VersionID) || (input.GenerationID != "" && !validSHA256Hex(input.GenerationID)) {
		return out, errors.New("email PDF requires an exact version and decoded generation")
	}
	if input.Paper == "" {
		input.Paper = "A4"
	}
	if input.Paper != "A4" && input.Paper != "Letter" {
		return out, errors.New("email PDF paper must be A4 or Letter")
	}
	if err := c.do(ctx, http.MethodPost, "/api/v1/email-pdfs", nil, input, &out); err != nil {
		return out, err
	}
	if out.VersionID != input.VersionID || !validSHA256Hex(out.ProfileFingerprint) || !validSHA256Hex(out.JobID) {
		return document.EmailPDFJob{}, integrityErrorf("email PDF job identity disagrees with request")
	}
	return out, nil
}

func (c *Client) EmailPDFJobState(ctx context.Context, jobID string) (string, error) {
	if !validSHA256Hex(jobID) {
		return "", errors.New("email PDF requires a job fingerprint")
	}
	var out struct {
		State string `json:"state"`
	}
	if err := c.do(ctx, http.MethodGet, "/api/v1/email-pdf-jobs/"+url.PathEscape(jobID), nil, nil, &out); err != nil {
		return "", err
	}
	switch out.State {
	case "queued", "running", "retry_wait", "operator_required", "failed", "completed":
	default:
		return "", integrityErrorf("email PDF job has an invalid state")
	}
	return out.State, nil
}

func emailPDFPath(versionID, profile string) string {
	return "/api/v1/email-pdfs/" + url.PathEscape(versionID) + "/" + url.PathEscape(profile)
}

func (c *Client) EmailPDFReceipt(ctx context.Context, versionID, profile string) (document.EmailPDFReceiptV1, error) {
	var out document.EmailPDFReceiptV1
	if !validUUIDv4(versionID) || !validSHA256Hex(profile) {
		return out, errors.New("email PDF requires exact version and profile")
	}
	if err := c.do(ctx, http.MethodGet, emailPDFPath(versionID, profile), nil, nil, &out); err != nil {
		return out, err
	}
	if out.Source.VersionID != versionID || out.ProfileFingerprint != profile || !validSHA256Hex(out.AttachmentID) || !validSHA256Hex(out.BuildID) || document.ValidateEmailPDFBinding(out.Binding) != nil || !validSHA256Hex(out.Output.PDFSHA256) || out.Output.PDFSize < 1 || out.Output.PDFSize > emailpdf.MaxPDFBytes || out.Output.Pages < 1 || out.Output.Pages > emailpdf.MaxPDFPages {
		return document.EmailPDFReceiptV1{}, integrityErrorf("email PDF retained receipt is inconsistent")
	}
	metadata, err := c.EmailMetadataGeneration(ctx, versionID, out.Binding.GenerationID)
	if err != nil {
		return document.EmailPDFReceiptV1{}, err
	}
	if metadata.Checksum != out.Binding.GenerationChecksum || metadata.Version.NodeID != out.Source.NodeID || metadata.Version.BlobHash != out.Source.SHA256 || metadata.Version.Size != out.Source.Size {
		return document.EmailPDFReceiptV1{}, integrityErrorf("email PDF source and decoded generation disagree")
	}
	return out, nil
}

// DownloadEmailPDF verifies the exact typed receipt and complete bytes; it
// never represents the derivative as an original content version.
func (c *Client) DownloadEmailPDF(ctx context.Context, receipt document.EmailPDFReceiptV1) ([]byte, error) {
	verified, err := c.EmailPDFReceipt(ctx, receipt.Source.VersionID, receipt.ProfileFingerprint)
	if err != nil {
		return nil, err
	}
	if verified != receipt {
		return nil, integrityErrorf("email PDF receipt changed")
	}
	req, err := c.emailRequest(ctx, http.MethodGet, emailPDFPath(receipt.Source.VersionID, receipt.ProfileFingerprint)+"/content", nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.hc.Do(req)
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode != http.StatusOK {
		return nil, decodeError(resp)
	}
	b, err := io.ReadAll(io.LimitReader(resp.Body, receipt.Output.PDFSize+1))
	if err != nil {
		return nil, err
	}
	sum := sha256.Sum256(b)
	pages, parseErr := emailpdf.VerifyPDF(b)
	if int64(len(b)) != receipt.Output.PDFSize || hex.EncodeToString(sum[:]) != receipt.Output.PDFSHA256 || parseErr != nil || pages != receipt.Output.Pages {
		return nil, integrityErrorf("email PDF bytes disagree with verified receipt")
	}
	return b, nil
}
